modulus/utils/zenith_angle.py

```diff
@@ -101,6 +105,197 @@ def cos_zenith_angle_from_timestamp(
     return _star_cos_zenith(julian_centuries, lon_rad, lat_rad)
 
 
+def irradiance(
+    t,
+    S0=1361,
+    e=0.0167,
+    perihelion_longitude=282.895,
+    mean_tropical_year=365.2422,
+    newton_iterations: int = 3,
+):
+    """The flux of solar energy in W/m2 towards Earth
+
+    The default orbital parameters are set to 2000 values.
+    Over the period of 1900-2100 this will result in an error of at most 0.02%,
+    so can be neglected for many applications.
+
+    Args:
+        t: linux timestamp
+        S0: the solar constant in W/m2. This is the mean irradiance received by
+            earth over a year.
+        e: the eccentricity of earths elliptical orbit
+        perihelion_longitude: spatial angle from moving vernal equinox to perihelion with Sun as angle vertex.
+            Perihelion is moment when earth is closest to sun. vernal equinox is
+            the longitude when the Earth crosses the equator from South to North.
+        newton_iterations: number of iterations for newton solver for elliptic anomaly
+
+
+    Notes:
+
+        TISR can be computed from Berger's formulas:
+
+            Berger, A. (1978). Long-Term Variations of Daily Insolation and
+            Quaternary Climatic Changes. Journal of the Atmospheric Sciences,
+            35(12), 2362–2367.
+            https://doi.org/10.1175/1520-0469(1978)035<2362:LTVODI>2.0.CO;2
+
+        NASA Example computing the orbital parameters: https://data.giss.nasa.gov/modelE/ar5plots/srorbpar.html. From 1900-2100 these are the ranges::
+            Orbital Parmameters
+
+                                                Long. of
+            Year     Eccentri    Obliquity    Perihel.
+            (A.D.)      city      (degrees)    (degrees)
+            ------    --------    ---------    --------
+                1900   0.016744      23.4528      281.183
+                2000   0.016704      23.4398      282.895
+                2100   0.016663      23.4268      284.609
+
+    """
+    seconds_per_solar_day = 86400
+    mean_tropical_year = mean_tropical_year * seconds_per_solar_day
+
+    year_2000_equinox = datetime.datetime(2000, 3, 20, 7, 35, tzinfo=pytz.utc)
+
+    # from appendix of Berger 1978
+    M = (t - year_2000_equinox.timestamp()) % mean_tropical_year
+    M = M / mean_tropical_year * 2 * np.pi
+    M -= np.deg2rad(perihelion_longitude)
+
+    # to get the elliptic anomaly E from the "mean anomaly" M
+    # use eq. 6.37
+    # https://link.springer.com/book/10.1007/978-3-662-53045-0)
+    # r / a = (1 - e cos E )
+    # E - e sin(E) = M
+    def f(E):
+        return E - e * np.sin(E) - M
+
+    def fp(E):
+        return 1 - e * np.cos(E)
+
+    # newton iterations
+    # initial guess
+    E = M / (1 - e)
```

Comment:
Optional, but might be an enhancement:

This initial condition is a good one in the limit `E -> 0`, but a better asymptotic expansion is to do `(E-M) -> 0`. If you do that, the first order solution ends up being

```math
E = M + \frac{e \sin M}{1 - e \cos M}
```

And it is satisfying that in the limit that `M -> 0`, you recover the initial guess that you have 😄